### PR TITLE
Color Palette

### PR DESCRIPTION
**Pull request by Skalii**

### CODE REVIEW

File: Sources/Plugins/SRTMPlugin/TerrainMode.swift [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+//  TerrainMode.swift
+//  OsmAnd Maps
+//
+//  Created by Skalii on 19.06.2024.
+//  Copyright © 2024 OsmAnd. All rights reserved.
+//
+
+import Foundation
+
+@objc
```
Comment (tigrim) [inline Sources/Plugins/SRTMPlugin/TerrainMode.swift:11]: remove

File: Sources/Plugins/SRTMPlugin/TerrainMode.swift [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    static let heightPrefix = "height_"
+
+    let type: TerrainType
+
+    private let minZoom: OACommonInteger
+    private let maxZoom: OACommonInteger
+    private let transparency: OACommonInteger
+    private let key: String
+
+    static var values: [TerrainMode] {
```
Comment (tigrim) [inline Sources/Plugins/SRTMPlugin/TerrainMode.swift:51]: static var on top

File: Sources/Plugins/SRTMPlugin/TerrainMode.swift [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        if let terrainModes {
```
Comment (tigrim) [inline Sources/Plugins/SRTMPlugin/TerrainMode.swift:52]: guard let terrainModes

File: Sources/Plugins/SRTMPlugin/TerrainMode.swift [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        self.type = type
+        self.translateName = translateName
+
+        let settings = OAAppSettings.sharedManager()!
+        minZoom = settings.registerIntPreference(key + "_min_zoom", defValue: 3).makeProfile()
+        maxZoom = settings.registerIntPreference(key + "_max_zoom", defValue: 17).makeProfile()
+        transparency = settings.registerIntPreference(key + "_transparency", defValue: type == .hillshade ? 100 : 80).makeProfile()
+    }
+
+    static func reloadTerrainModes() {
```
Comment (tigrim) [inline Sources/Plugins/SRTMPlugin/TerrainMode.swift:74]: private?

File: Sources/Plugins/SRTMPlugin/TerrainMode.swift [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        modes.append(TerrainMode(defaultKey, type: .hillshade, translateName: localizedString("shared_string_hillshade")))
+        modes.append(TerrainMode(defaultKey, type: .slope, translateName: localizedString("shared_string_slope")))
+        
+        if let dir = OsmAndApp.swiftInstance().colorsPalettePath,
+           let files = try? FileManager.default.contentsOfDirectory(atPath: dir) {
+            for file in files {
+                guard file.hasSuffix(TXT_EXT) else { continue }
+
+                let nm = file
+                if nm.hasPrefix(hillshadePrefix) {
```
Comment (tigrim) [inline Sources/Plugins/SRTMPlugin/TerrainMode.swift:85]: It seems we can simplify by extracting the implementation into a separate method and iterating through a dictionary like [.hillshade: hillshadePrefix, ...]

File: Sources/Plugins/SRTMPlugin/TerrainMode.swift [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+                        modes.append(TerrainMode(key, type: .height, translateName: name))
+                    }
+                }
+            }
+        }
+        terrainModes = modes
+    }
+
+    static func getMode(_ type: TerrainType, keyName: String) -> TerrainMode? {
+        if let terrainModes {
```
Comment (tigrim) [inline Sources/Plugins/SRTMPlugin/TerrainMode.swift:110]: guard

File: Sources/Plugins/SRTMPlugin/TerrainMode.swift [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+            for mode in terrainModes {
+                if mode.type == type && mode.getKeyName() == keyName {
```
Comment (tigrim) [inline Sources/Plugins/SRTMPlugin/TerrainMode.swift:112]: where

File: Sources/Plugins/SRTMPlugin/TerrainMode.swift [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+                    return mode
+                }
+            }
+        }
+        return nil
+    }
+
+    static func getDefaultMode(_ type: TerrainType) -> TerrainMode? {
+        if let terrainModes {
```
Comment (tigrim) [inline Sources/Plugins/SRTMPlugin/TerrainMode.swift:121]: up

File: Sources/Plugins/SRTMPlugin/TerrainMode.swift [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+        type == .hillshade
+    }
+
+    func isSlope() -> Bool {
+        type == .slope
+    }
+
+    func getMainFile() -> String {
+        let prefix: String
+        if type == .height {
```
Comment (tigrim) [inline Sources/Plugins/SRTMPlugin/TerrainMode.swift:150]: use switch

File: Sources/Plugins/SRTMPlugin/TerrainMode.swift [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+    func getMaxZoom() -> Int32 {
+        maxZoom.get()
+    }
+
+    func getDefaultDescription() -> String {
+        translateName
+    }
+
+    func getDescription() -> String {
+        if type == .hillshade {
```
Comment (tigrim) [inline Sources/Plugins/SRTMPlugin/TerrainMode.swift:223]: switch

File: Sources/Router/Colorization/ColorPalette.swift [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+//  ColorPalette.swift
+//  OsmAnd
+//
+//  Created by Skalii on 26.06.2024.
+//  Copyright © 2024 OsmAnd. All rights reserved.
+//
+
+import Foundation
+
+@objc
```
Comment (tigrim) [inline Sources/Router/Colorization/ColorPalette.swift:11]: ?

File: Sources/Router/Colorization/ColorPalette.swift [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+@objcMembers
+final class ColorPalette: NSObject {
+
+    @objc
```
Comment (tigrim) [inline Sources/Router/Colorization/ColorPalette.swift:15]: ?

File: Sources/Router/Colorization/ColorPalette.swift [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+            if let c = c {
+                palette.colorValues.append(c)
+            }
+        }
+        palette.sortPalette()
+        return palette
+    }
+    
+    static func parseColorPalette(from filePath: String) throws -> ColorPalette {
+        var comments = [String]()
```
Comment (tigrim) [inline Sources/Router/Colorization/ColorPalette.swift:138]: Why do u need inout var comments?

File: Sources/Router/Colorization/ColorPaletteHelper.swift [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+//  ColorPaletteHelper.swift
+//  OsmAnd Maps
+//
+//  Created by Skalii on 02.07.2024.
+//  Copyright © 2024 OsmAnd. All rights reserved.
+//
+
+import Foundation
+
+@objc
```
Comment (tigrim) [inline Sources/Router/Colorization/ColorPaletteHelper.swift:11]: up

File: Sources/Router/Colorization/ColorPaletteHelper.swift [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+        if !isTerrainType {
+            return getColorizationTypePallets(EOAColorizationType(rawValue: gradientType)!)
+        } else {
+            return getTerrainModePallets(TerrainMode.TerrainType(rawValue: Int32(gradientType))!)
+        }
+    }
+
+    func requireGradientColorPaletteSync(_ colorizationType: EOAColorizationType, gradientPaletteName: String) -> ColorPalette {
+        let colorPalette = getGradientColorPaletteSync(colorizationType, gradientPaletteName: gradientPaletteName)
+        return isValidPalette(colorPalette) ? colorPalette! : OARouteColorize.getDefaultPalette(colorizationType)
```
Comment (tigrim) [inline Sources/Router/Colorization/ColorPaletteHelper.swift:34]: force unwrap

File: Sources/Router/Colorization/ColorPaletteHelper.swift [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    func getGradientColorPaletteSync(_ colorizationType: EOAColorizationType, gradientPaletteName: String) -> ColorPalette? {
+        return getGradientColorPaletteSync(colorizationType, gradientPaletteName: gradientPaletteName, refresh: false)
```
Comment (tigrim) [inline Sources/Router/Colorization/ColorPaletteHelper.swift:38]: return

File: Sources/Router/Colorization/ColorPaletteHelper.swift [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    func getGradientColorPaletteSync(_ colorizationType: EOAColorizationType, gradientPaletteName: String, refresh: Bool) -> ColorPalette? {
+        let colorPaletteFileName = "route_\(getColorizationTypeName(colorizationType))_\(gradientPaletteName)\(TXT_EXT)"
+        return getGradientColorPalette(colorPaletteFileName, refresh: refresh)
```
Comment (tigrim) [inline Sources/Router/Colorization/ColorPaletteHelper.swift:43]: up

File: Sources/Router/Colorization/ColorPaletteHelper.swift [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    func getGradientColorPaletteSyncWithModeKey(_ modeKey: String) -> ColorPalette? {
+        return getGradientColorPalette(modeKey)
```
Comment (tigrim) [inline Sources/Router/Colorization/ColorPaletteHelper.swift:47]: up

File: Sources/Router/Colorization/ColorPaletteHelper.swift [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    func getGradientColorPalette(_ colorPaletteFileName: String) -> ColorPalette? {
+        return getGradientColorPalette(colorPaletteFileName, refresh: false)
```
Comment (tigrim) [inline Sources/Router/Colorization/ColorPaletteHelper.swift:51]: up

File: Sources/Router/Colorization/ColorPaletteHelper.swift [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+            return cachedPalette
+        }
+        let filePath = getColorPaletteDir().appendingPathComponent(colorPaletteFileName)
+        if FileManager.default.fileExists(atPath: filePath) {
+            do {
+                let colorPalette = try ColorPalette.parseColorPalette(from: filePath)
+                cachedColorPalette[colorPaletteFileName] = colorPalette
+                return colorPalette
+            } catch {
+                print("Error reading color file: \(error)")
```
Comment (tigrim) [inline Sources/Router/Colorization/ColorPaletteHelper.swift:65]: debugPrint or NSLog

File: Sources/Router/Colorization/ColorPaletteHelper.swift [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+                    let colorPalleteName = fileName.replacingOccurrences(of: colorTypePrefix, with: "").replacingOccurrences(of: TXT_EXT, with: "")
+                    if let colorPalette = getGradientColorPalette(fileName) {
+                        let fileAttributes = try FileManager.default.attributesOfItem(atPath: fileName)
+                        let modificationDate = fileAttributes[.modificationDate] as! Date
+                        colorPalettes[colorPalleteName] = [colorPalette, modificationDate.timeIntervalSince1970]
+                    }
+                }
+            }
+        } catch {
+            print("Error reading color palette directory: \(error)")
```
Comment (tigrim) [inline Sources/Router/Colorization/ColorPaletteHelper.swift:88]: up

File: Sources/Router/Colorization/ColorPaletteHelper.swift [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+        }
+        
+        return colorPalettes
+    }
+
+    private func getTerrainModePallets(_ type: TerrainMode.TerrainType) -> [String: [Any]] {
+        var colorPalettes: [String: [Any]] = [:]
+        let modes = TerrainMode.values
+        for mode in modes {
```
Comment (tigrim) [inline Sources/Router/Colorization/ColorPaletteHelper.swift:97]: use without temp var modes

File: Sources/Router/Colorization/ColorPaletteHelper.swift [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+                    let modificationDate = fileAttributes?[.modificationDate] as? Date
+                    colorPalettes[mode.getKeyName()] = [colorPalette, modificationDate?.timeIntervalSince1970 ?? 0]
+                }
+            }
+        }
+        return colorPalettes
+    }
+
+    private func isValidPalette(_ palette: ColorPalette?) -> Bool {
+        return palette != nil && palette!.colorValues.count >= 2
```
Comment (tigrim) [inline Sources/Router/Colorization/ColorPaletteHelper.swift:112]: force unwrap

File: Sources/Plugins/SRTMPlugin/TerrainMode.swift (hunk 6 above)
Comment (tigrim) [reply Sources/Plugins/SRTMPlugin/TerrainMode.swift:74]: +

File: Sources/Plugins/SRTMPlugin/TerrainMode.swift (hunk 8 above)
Comment (tigrim) [reply Sources/Plugins/SRTMPlugin/TerrainMode.swift:83]: +

File: Sources/Controllers/Map/Layers/OAGPXLayer.mm [hunk 24]
```diff
@@ -244,12 +244,24 @@ - (void)configureCachedWallColorsFor:(EOAGradientScaleType)type
...
+                                            analysis:analysis ?: [doc getAnalysis:0]
+                                                type:[type toColorizationType]
+                                             palette:palette
+                                     maxProfileSpeed:0];
+        _cachedWallColors[key].clear();
+        if (routeColorize)
+        {
+            for (OARouteColorizationPoint *colorizationPoint in [routeColorize getResult])
+            {
+                _cachedWallColors[key].append(OsmAnd::ColorARGB(colorizationPoint.color));
```
Comment (tigrim) [inline Sources/Controllers/Map/Layers/OAGPXLayer.mm:262]: Why not just return QList of colors immediately?
 _cachedWallColors[key] =  [routeColorize colors];

File: Sources/Controllers/Map/Layers/OAGPXLayer.mm [hunk 25]
```diff
@@ -305,21 +317,49 @@ - (void) refreshGpxTracks:(BOOL)refreshColors
...
+                                                            type:[type toColorizationType]
+                                                         palette:palette
+                                                 maxProfileSpeed:0];
+                    _cachedColors[key].clear();
+                    if (routeColorize)
                     {
-                        _cachedColors[key].append(OsmAnd::ColorARGB(colorizationPoint.color));
+                        for (OARouteColorizationPoint *colorizationPoint in [routeColorize getResult])
+                        {
+                            _cachedColors[key].append(OsmAnd::ColorARGB(colorizationPoint.color));
```
Comment (tigrim) [inline Sources/Controllers/Map/Layers/OAGPXLayer.mm:355]: up

**Review by tigrim — APPROVED** [review]

File: Sources/Helpers/MigrationManager.swift [hunk 26]
```diff
@@ -381,6 +385,55 @@ final class MigrationManager: NSObject {
         }
     }
 
+    private func changeTerrainSettingsMigration1() {
+        if let plugin = OAPluginsHelper.getPlugin(OASRTMPlugin.self) as? OASRTMPlugin {
```
Comment (alex-osm) [inline Sources/Helpers/MigrationManager.swift:389]: Shouldn't the migration run even if the plugin is currently disabled?

File: Sources/Plugins/SRTMPlugin/OASRTMPlugin.h [hunk 27]
```diff
@@ -8,14 +8,26 @@
...
+static NSInteger const terrainMinSupportedZoom = 4;
+static NSInteger const terrainMaxSupportedZoom = 19;
+
+@class OACommonBoolean, TerrainMode;
 
 @interface OASRTMPlugin : OAPlugin
 
 @property (nonatomic) OACommonBoolean *enable3DMaps;
+@property (nonatomic) OACommonBoolean *terrain;
+@property (nonatomic) OACommonString *terrainModeType;
```
Comment (alex-osm) [inline Sources/Plugins/SRTMPlugin/OASRTMPlugin.h:20]: Better naming needed: enable3dMapsPref, terrainEnabledPref, terrainModeTypePref

File: Sources/Map/OATerrainLayer.mm [hunk 28]
```diff
@@ -76,7 +77,8 @@ - (instancetype) init:(EOATerrainLayerType)terrainType
 - (void) onTerrainResourcesChanged
 {
     [self initCollection];
-    [[OsmAndApp instance].data.terrainChangeObservable notifyEvent];
+    [[NSNotificationCenter defaultCenter] postNotificationName:kNotificationSetProfileSetting
```
Comment (alex-osm) [inline Sources/Map/OATerrainLayer.mm:80]: The kNotificationSetProfileSetting should be posted by OAAppSettings only (as it is done now)!!!

File: Sources/OsmEdit/OsmData/OAOsmMapUtils.h [hunk 29]
```diff
@@ -13,12 +13,17 @@
...
 NS_ASSUME_NONNULL_BEGIN
 
-@class OAWay;
+@class OAWay, OANode;
 
 @interface OAOsmMapUtils : NSObject
 
 +(CLLocationCoordinate2D)getWeightCenterForWay:(OAWay *)way;
 
++ (void)simplifyDouglasPeucker:(NSArray<OANode *> *)nodes
```
Comment (alex-osm) [inline Sources/OsmEdit/OsmData/OAOsmMapUtils.h:22]: It is not clear why do you need it in iOS?

File: Sources/Router/Colorization/OARouteColorize.mm [hunk 30]
```diff
@@ -245,40 +245,126 @@ - (void) correctElevations:(NSMutableArray<NSNumber *> *)elevations
...
+    NSArray<OARouteColorizationPoint *> *result = [self getResult];
+    QList<OsmAnd::FColorARGB> qResult;
+    for (OARouteColorizationPoint *colorizationPoint in result)
+    {
+        qResult.append(OsmAnd::ColorARGB(colorizationPoint.color));
+    }
+    return qResult;
+}
+
+- (NSArray<OARouteColorizationPoint *> *)getSimplifiedResult:(NSInteger)simplificationZoom
```
Comment (alex-osm) [inline Sources/Router/Colorization/OARouteColorize.mm:270]: It looks like `getSimplifiedResult`, `simplify` and perhaps other new methods are redundant. Why do we need them?

File: Sources/Plugins/ExternalSensorsPlugin/OAExternalSensorsPlugin.mm [hunk 31]
```diff
@@ -41,10 +41,10 @@ - (instancetype)init
...
 
-        _speedSensorWriteToTrackDeviceID = [self registerStringPreference:@"speed_sensor_write_to_track_device" defValue:@""];
-        _cadenceSensorWriteToTrackDeviceID = [self registerStringPreference:@"cadence_sensor_write_to_track_device" defValue:@""];
-        _heartSensorWriteToTrackDeviceID = [self registerStringPreference:@"heart_rate_sensor_write_to_track_device" defValue:@""];
-        _temperatureSensorWriteToTrackDeviceID = [self registerStringPreference:@"temperature_sensor_write_to_track_device" defValue:@""];
+        _speedSensorWriteToTrackDeviceID = [[self registerStringPreference:@"speed_sensor_write_to_track_device" defValue:@""] makeProfile];
+        _cadenceSensorWriteToTrackDeviceID = [[self registerStringPreference:@"cadence_sensor_write_to_track_device" defValue:@""] makeProfile];
+        _heartSensorWriteToTrackDeviceID = [[self registerStringPreference:@"heart_rate_sensor_write_to_track_device" defValue:@""] makeProfile];
+        _temperatureSensorWriteToTrackDeviceID = [[self registerStringPreference:@"temperature_sensor_write_to_track_device" defValue:@""] makeProfile];
```
Comment (alex-osm) [inline Sources/Plugins/ExternalSensorsPlugin/OAExternalSensorsPlugin.mm:47]: Is migration needed?

File: Sources/Plugins/SRTMPlugin/OASRTMPlugin.mm [hunk 32]
```diff
@@ -17,31 +17,37 @@
...
 #import "OALinks.h"
+#import "OsmAnd_Maps-Swift.h"
 
-#define PLUGIN_ID kInAppId_Addon_Srtm
-
-#define kEnable3dMaps @"enable_3d_maps"
+static NSString * const PLUGIN_ID = kInAppId_Addon_Srtm;
+static NSString * const kEnable3dMaps = @"enable_3d_maps";
+static NSString * const kTerrainMode = @"terrain_mode";
+static NSString * const kTerrain = @"terrain_layer";
```
Comment (alex-osm) [inline Sources/Plugins/SRTMPlugin/OASRTMPlugin.mm:25]: kEnable3dMapsPrefName, kTerrainModePrefName, kTerrainEnabledPrefName

File: Sources/Plugins/SRTMPlugin/TerrainMode.swift [hunk 33]
```diff
@@ -0,0 +1,237 @@
...
+        return terrainModes
+    }
+
+    private static var terrainModes: [TerrainMode]?
+
+    let type: TerrainType
+
+    private let minZoom: OACommonInteger
+    private let maxZoom: OACommonInteger
+    private let transparency: OACommonInteger
```
Comment (alex-osm) [inline Sources/Plugins/SRTMPlugin/TerrainMode.swift:57]: minZoomPref, maxZoomPref, transparencyPref

**Pull request merged**